src/Hedgehog/Plutus/Tx.hs

```diff
@@ -106,11 +128,116 @@ data ScriptPurpose
+  -- as every user at once in most cases
+  -- so we should probably make this an argument to balance
+  (valueIn, valueOut) <- getValueInAndOut context tx
+  -- TODO we may want to change this to something like
+  -- Either FailReason (Tx 'Balanced)
+  -- to distinguish things like failure by tx lookup
+  -- and failure by not enough value at keys
+  let extra = posDif valueIn valueOut
+  let deficit = posDif valueOut valueIn
+  let alreadySpending =
+        Vector.fromList $
+          catMaybes
+            [ Map.lookup (txInRef txin) (Model.mockUtxos mock)
+            | txin <- Set.toList $ txInputs tx
+            ]
+  spendGen <- spend mock deficit balanceingKeys alreadySpending
+  pure $ do
+    Spend outRefs moreChange <- spendGen
+    payOutKey <- Gen.element (Vector.toList balanceingKeys)
+    let change = extra <> moreChange
+    let checked =
+          confirmBalanced context $
+            tx
+              <> mempty
+                { txInputs =
+                    Set.fromList $
+                      Vector.toList $
+                        (`TxIn` Nothing) <$> outRefs
+                , txOutputs =
+                    pure $
+                      TxOut
+                        { txOutAddress = Plutus.pubKeyHashAddress payOutKey
+                        , txOutValue = change
+                        , txOutDatum = Nothing
+                        , txOutReferenceScript = Nothing
+                        }
+                }
+    case checked of
+      Just tx -> pure tx
+      Nothing -> error "balanceTx produced an unbalanced tx"
+
+getBalance :: TxContext -> Tx b -> Maybe Value
+getBalance context tx = do
+  case protocol of
+    Model.AlonzoParams params -> do
+      Right utxo <- pure $ Class.toUtxo mempty networkId []
+      -- TODO what should the utxo be for this?
+      pure $
+        maryToPlutus $
+          evaluateTransactionBalance @Alonzo.Era
+            params
+            utxo
+            (const True) -- TODO is this right?
+            body
+    Model.BabbageParams params -> do
+      -- TODO is there a good way to not repeat all this?
+      Right utxo <- pure $ Class.toUtxo mempty networkId []
+      -- This can't be lifted to the outer doblock because `toUtxo` itself
+      -- not the return value is polymorphic so it needs to be called twice
+      -- with the same arguments but a different infered type
+      pure $
+        maryToPlutus $
+          evaluateTransactionBalance @Babbage.Era
+            params
+            utxo
+            (const True)
+            body
+  where
+    networkId :: Network
+    networkId = Model.mockConfigNetworkId mockConfig
+
+    protocol :: Model.PParams
+    protocol = Model.mockConfigProtocol mockConfig
+
+    mockConfig :: Model.MockConfig
+    mockConfig = Model.mockConfig $ mockchain context
+
+    body :: Class.IsCardanoTx era => TxBody era
+    body = Class.getTxBody $ toLedgerTx context tx
+
+maryToPlutus :: MV.MaryValue era -> Plutus.Value
```

Comment:
It's in the back of my head that there's a library function for this.

Reply:
I wouldn't be surprised but I couldn't find one for the life of me. Relatedly do you know how to get hoogle working for this repo?